server/commands/list_objects.go

```diff
@@ -0,0 +1,234 @@
+			ObjectIds: objectIDs,
+		}, nil
+	case genericError := <-errChan:
+		return nil, genericError
+	}
+}
+
+func (q *ListObjectsQuery) ExecuteStreamed(ctx context.Context, req *openfgapb.StreamedListObjectsRequest, srv openfgapb.OpenFGAService_StreamedListObjectsServer) error {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return err
+	}
+
+	resultsChan := make(chan string, q.listObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.listObjectsDeadlineInSeconds != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, time.Second*time.Duration(q.listObjectsDeadlineInSeconds))
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, req.StoreId, req.AuthorizationModelId, req.Type, req.Relation, req.User, req.ContextualTuples, resultsChan, errChan, resolvedChan)
+	}()
+
+	for {
+		select {
+		case <-timeoutCtx.Done():
+			return serverErrors.NewInternalError("Timeout exceeded", timeoutCtx.Err())
+		case objectID, ok := <-resultsChan:
+			if !ok {
+				return nil //channel was closed
+			}
+			if err := srv.Send(&openfgapb.StreamedListObjectsResponse{
+				ObjectId: objectID,
+			}); err != nil {
+				return serverErrors.NewInternalError("", err)
+			}
+		case genericError, ok := <-errChan:
+			if ok {
+				return serverErrors.NewInternalError("", genericError)
+			}
+		}
+	}
+}
+
+func (q *ListObjectsQuery) getUniqueObjects(ctx context.Context, storeID, targetObjectType string, ctxTuples *openfgapb.ContextualTupleKeys) ([]string, error) {
+	uniqueObjects, err := q.datastore.ListObjectsByType(ctx, storage.ListObjectsFilter{
+		StoreID:    storeID,
+		ObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+
+	if ctxTuples != nil {
+		// Take the relevant objects (i.e. same type) from the contextual tuples and add them to the result
+		uniqueSet := make(map[string]bool, len(uniqueObjects))
+		for _, o := range uniqueObjects {
+			_, found := uniqueSet[o]
+			if !found {
+				uniqueSet[o] = true
+			}
+		}
+
+		for _, ctxTuple := range ctxTuples.TupleKeys {
+			objectType, _ := tuple.SplitObject(ctxTuple.Object)
+			if objectType != targetObjectType {
+				continue
+			}
+			_, found := uniqueSet[ctxTuple.Object]
+			if !found {
+				uniqueSet[ctxTuple.Object] = true
+				uniqueObjects = append(uniqueObjects, ctxTuple.Object)
+			}
+		}
+	}
+	return uniqueObjects, nil
+}
+
+func (q *ListObjectsQuery) validateInput(ctx context.Context, storeID string, targetObjectType string, authModelID string, relation string) error {
+	definition, err := q.datastore.ReadTypeDefinition(ctx, storeID, authModelID, targetObjectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return serverErrors.TypeNotFound(targetObjectType)
+		}
+		return err
+	}
+	_, ok := definition.Relations[relation]
```

Comment:
We have this same kind of validation elsewhere. Can we please refactor this into a shared component so that it can be used elsewhere?

We use it in `validateAndAuthenticateTuple` and others.

Reply:
I see only one other duplication, in `validateAndAuthenticateTuple` like you said, but the error returned for when the relation is invalid is different because in my PR i don't have a tuple to reference. I'm inclined to keep this as is

Reply:
This logic is overlapping though. You're only looking up the `*openfgapb.Userset` value for a specific `(objectType, relation)` pair. If the error messages turns a tuple in one and we only return a objectType error message, here then we should refactor to just return a generic error such as `fmt.Errorf("'%s' undefined for object type '%s'", relation, objectType)`. The code duplication isn't warranted just for a different error message.

This is also common enough functionality that will be useful elsewhere. Every Expand, Check, ListObjects, etc.. will start with validating a relation on a given objectType exists. So it should be shared.